### PR TITLE
Added support for Django 2.2 -- BOM-1042

### PR DESCRIPTION
**Pull request by mraarif**
Relevant JIRA issue can be found [here](https://openedx.atlassian.net/browse/BOM-1042).

- As per Django 2.2 release notes, to use SQLite as a backend we need to have migrations in apps referencing other apps with migrations [(source)](https://docs.djangoproject.com/en/3.0/releases/2.2/#miscellaneous)
- After adding the migration the tests for utils failed, I've updated mechanism how models are dynamically loaded based on the assumption that the Django app will have appropriate models in `models.py` file.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+- 3.5
 env:
-- TOXENV=master
+- TOXENV=django111
+- TOXENV=django20
+- TOXENV=django21
+- TOXENV=django22
 matrix:
   include:
   - python: 3.6
```
Comment (UsamaSadiq) [inline .travis.yml]: It should be `python: 3.5`.

File: tox.ini [hunk 2]
```diff
@@ -35,20 +35,19 @@ norecursedirs = .* docs requirements
 
 [testenv]
 deps =
-    django18: Django>=1.8,<1.9
     django111: Django>=1.11,<2.0
     django20: Django>=2.0,<2.1
+    django21: Django>=2.1,<2.2
+    django22: Django>=2.2,<2.3
```
Comment (UsamaSadiq) [inline tox.ini]: we need to use `django.txt` file for Django2.2 version as we did in other repos. Update `MakeFile` too for this change. For reference see https://github.com/edx/portal-designer/pull/84

File: requirements/test.txt [hunk 3]
```diff
@@ -4,57 +4,53 @@
...
 django-crum==0.7.5
 django-model-utils==3.2.0
-django-waffle==0.18.0
-djangorestframework-jwt==1.11.0
-djangorestframework==3.9.4
-edx-django-utils==2.0.2
-edx-drf-extensions==2.4.5
+django-waffle==0.20.0
+djangorestframework==3.11.0
+drf-jwt==1.15.1
```
Comment (UsamaSadiq) [inline requirements/test.txt]: There is a breaking change in `version>1.15.0`. See https://github.com/edx/portal-designer/pull/84 for reference. You'll need to add some installed applications and set the default to `JWT`.

File: .travis.yml (hunk 1 above)
Comment (UsamaSadiq) [reply .travis.yml]: You can also keep both version `3.6` and `3.5` for now, I think.

Comment (awais786) [inline .travis.yml]: You can use 3.5 instead of 3.6

File: edx_rbac/migrations/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
```
Comment (awais786) [inline edx_rbac/migrations/0001_initial.py]: whats a purpose of this migration ?

File: edx_rbac/utils.py [hunk 5]
```diff
@@ -114,8 +113,9 @@ def append_role_auth_claim(role_string, context=None):
             role_func = getattr(module, func_name)
         except (ImportError, AttributeError):
             # otherwise, assume that it's a django model
-            module = apps.get_model(module_name, func_name)
-            role_func = module.get_assignments
+            models = importlib.import_module('{0}.models'.format(module_name))
```
Comment (awais786) [inline edx_rbac/utils.py]: whats a purpose of these changes plz add some explanation.

File: tests/test_utils.py [hunk 6]
```diff
@@ -239,9 +239,8 @@ class TestUtilsWithDatabaseRequirements(TestCase):
 
     def setUp(self):
         super(TestUtilsWithDatabaseRequirements, self).setUp()
-        self.user = User.objects.create(username='test_user', password='pw')
-        self.role = ConcreteUserRole(name='coupon-manager')
-        self.role.save()
+        self.user = User.objects.get_or_create(username='test_user', password='pw')[0]
```
Comment (awais786) [inline tests/test_utils.py]:
```
get_or_create returns object and bool value so it will break in case of already created user.
```

Comment (awais786) [inline tests/test_utils.py]:
```
get_or_create returns tuple of (object and bool) value. Please double check

File: tox.ini (hunk 2 above)
Comment (mraarif) [reply tox.ini]: @jmbowman against which version of Django do we want to deploy this package? `1.11` or `2.2`?

File: edx_rbac/migrations/0001_initial.py (hunk 4 above)
Comment (mraarif) [reply edx_rbac/migrations/0001_initial.py]: updated the description.

File: edx_rbac/utils.py (hunk 5 above)
Comment (mraarif) [reply edx_rbac/utils.py]: description updated.

File: requirements/test.txt (hunk 3 above)
Comment (mraarif) [reply requirements/test.txt]: done

File: .travis.yml (hunk 1 above)
Comment (mraarif) [reply .travis.yml]: done

File: edx_rbac/migrations/0001_initial.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    initial = True
+
+    dependencies = [
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='ConcreteUserRole',
```
Comment (jmbowman) [inline edx_rbac/migrations/0001_initial.py]: This migration shouldn't be in the `edx_rbac` app, since these models are only needed for tests.  You should treat the `tests` directory where these models are defined as a Django app, and put this migrations directory in it instead.  You should just need to add `tests` to `INSTALLED_APPS` in the test settings and define an [AppConfig](https://docs.djangoproject.com/en/2.2/ref/applications/#configuring-applications).

File: edx_rbac/utils.py (hunk 5 above)
Comment (jmbowman) [reply edx_rbac/utils.py]: You may not need this change after moving the new migrations to the `tests` directory.

File: requirements/base.txt [hunk 8]
```diff
@@ -6,29 +6,30 @@
...
-djangorestframework-jwt==1.11.0  # via edx-drf-extensions
-djangorestframework==3.9.4  # via edx-drf-extensions, rest-condition
-edx-django-utils==2.0.2   # via edx-drf-extensions
-edx-drf-extensions==2.4.5
+django-crum==0.7.5        # via -r requirements/base.in
+django-model-utils==3.2.0  # via -r requirements/base.in
+django-waffle==0.20.0     # via edx-django-utils, edx-drf-extensions
+django==2.2.11            # via -r requirements/base.in, django-crum, django-model-utils, djangorestframework, drf-jwt, edx-django-utils, edx-drf-extensions, rest-condition
+djangorestframework==3.11.0  # via drf-jwt, edx-drf-extensions, rest-condition
+drf-jwt==1.15.1           # via edx-drf-extensions
```
[the remark this answers is not in the document]
Comment (mraarif) [reply requirements/base.txt]: this would depend on `edx-drf-extensions` release for BOM-1410 changes I guess

File: requirements/base.txt [hunk 9]
```diff
@@ -6,29 +6,30 @@
...
-django-model-utils==3.2.0
-django-waffle==0.18.0     # via edx-django-utils, edx-drf-extensions
-django==1.11.26
-djangorestframework-jwt==1.11.0  # via edx-drf-extensions
-djangorestframework==3.9.4  # via edx-drf-extensions, rest-condition
-edx-django-utils==2.0.2   # via edx-drf-extensions
-edx-drf-extensions==2.4.5
+django-crum==0.7.5        # via -r requirements/base.in
+django-model-utils==3.2.0  # via -r requirements/base.in
+django-waffle==0.20.0     # via edx-django-utils, edx-drf-extensions
```
[the remark this answers is not in the document]
Comment (mraarif) [reply requirements/base.txt]: `django-waffle` isn't a direct dependency, it is required by `edx-drf-extensions`, maybe we can constrain its version in next release for `edx-drf-extensions`

File: edx_rbac/migrations/0001_initial.py (hunk 7 above)
Comment (mraarif) [reply edx_rbac/migrations/0001_initial.py]: We might need that because the current build is failing (after making the suggested changes) against Django 2.2, as `edx_rbac` is a Django app and it's referencing other Django models (with apps containing migrations), or we can just move models to `tests` app

Comment (jmbowman) [reply edx_rbac/migrations/0001_initial.py]: The only models in edx_rbac itself are abstract and have no associated tables.  The ones causing the tests to fail are the concrete subclasses of them in `tests/models.py`, so those are the ones running afoul of the rule you mentioned; that can be solved by turning `tests` into its own Django app.

File: test_settings.py [hunk 10]
```diff
@@ -32,7 +32,7 @@ def root(*args):
     'django.contrib.auth',
     'django.contrib.contenttypes',
     'edx_rbac',
-    'tests',
```
Comment (feanil) [inline test_settings.py]: Why did you rename this app?

File: edx_rbac/migrations/0001_initial.py (hunk 7 above)
Comment (feanil) [reply edx_rbac/migrations/0001_initial.py]: I quickly did a test to see if we really needed this migration by removing it. All tests pass even if this file doesn't exist.  I don't think it's necessary.  If you're still seeing an error when this file doesn't exist can you give me some steps to re-produce?

File: test_settings.py (hunk 10 above)
Comment (mraarif) [reply test_settings.py]: It wasn't a Django app before, I had to create the Django app with a different name to generate migrations in the same app, as mentioned in the PR description (it works locally without adding migrations) it fails on Django 2.2 in Travis if the app doesn't have migrations with Sqlite.

Comment (feanil) [reply test_settings.py]: @mraarif where/how are you re-producing this error? I don't see a jenkins build with it and I couldn't reproduce it locally.  I ran into a different discussion related error when I tried to run pytset in edx-platform with this change.

Comment (jmbowman) [reply test_settings.py]: https://travis-ci.org/github/edx/edx-rbac/builds/663497804

**Comment by feanil** [comment]
@mraarif I'm still having trouble reproducing this.  I made this branch https://github.com/edx/edx-rbac/tree/feanil/django22 with only a few changes.  I then updated edx-pltaform locally to use my new version of edx-rbac.

I ran the following:
```
tox -r -e py35-django22 -- pytest openedx/core/djangoapps/system_wide_roles/tests/test_models.py
```

Which should should be exercising the edx-rbac code.  But those test passed in django 2.2 for me locally.

**Comment by mraarif** [comment]
@feanil the travis [worker](https://travis-ci.org/github/edx/edx-rbac/jobs/666364662?utm_medium=notification&utm_source=github_status) for Django 2.2 in your PR is actually installing Django 1.11

and in one the previous [runs](https://travis-ci.org/github/edx/edx-rbac/jobs/663497808) under Django 2.2 worker (where it installed django2.2) in this PR it failed.
I think Django installation in your PR gets overridden by one of the requirement files.

File: edx_rbac/migrations/0001_initial.py (hunk 4 above)
Comment (feanil) [inline edx_rbac/migrations/0001_initial.py]: This migration should be moved into the `test_app` django app.  Since we only need it for testing.  You can see an example of how this was managed in edx-when: https://github.com/edx/edx-when/tree/master/tests/test_models_app/migrations

If you leave it in the edx_rback folder, it will be seen as a migration for this app when we install it in edx-platform and that will cause all sorts of upstream issues.

File: edx_rbac/utils.py (hunk 5 above)
Comment (feanil) [reply edx_rbac/utils.py]: +1 to what Jeremy said.  Once the migration is moved into the `test_app/migrations` directory. I think you should be able to undo the changes in this file.

File: requirements/django.txt [hunk 11]
```diff
@@ -0,0 +1 @@
+django==2.2.11            # via -r requirements/base.in, django-crum, django-model-utils, djangorestframework, drf-jwt, edx-django-utils, edx-drf-extensions, rest-condition
```
Comment (feanil) [inline requirements/django.txt]: Can you explain why you need this file?  When I removed the constraint in `base.in`  then the tox setup worked fine for me without this addition.

Comment (jmbowman) [reply requirements/django.txt]: Mainly to be precise about the exact version being tested.  This also keeps development environments up to date as new Django point releases come out.  In services it's more important, since it enforces that we've tested with the exact version being deployed.

**Comment by mraarif** [comment]
@feanil after moving the migrations to test app (following edx-when), tests are still failing under Django 2.2

I'm not sure how can we work around this as long as `edx_rbac` has models and doesn't have migrations.

**Comment by jmbowman** [comment]
The remaining problems with the models move were mostly caused by changes in `models.py` and `utils.py` left over from the first attempt at fixing the Django 2.2 tests; the AppConfig for the new Django app also wasn't being found by the test settings.  With those fixed, the `tests` directory itself can be used as the new Django app.
